engine/src/cubos/engine/gizmos/plugin.cpp

```diff
@@ -1,29 +1,202 @@
+        {
+            auto [localToWorld, camera] = *components;
+            auto v = glm::inverse(localToWorld->mat);
+            auto p =
+                glm::perspective(glm::radians(camera->fovY), float(viewports[i].size.x) / float(viewports[i].size.y),
+                                 camera->zNear, camera->zFar);
+            mvps[cameraCount] = p * v;
+            cameraCount++;
+        }
+    }
+
+    std::vector<std::size_t> toRemove;
+    for (std::size_t i = 0; i < gizmos.worldGizmos.size(); i++)
+    {
+        auto& gizmo = *(gizmos.worldGizmos[i]);
+
+        for (int i = 0; i < cameraCount; ++i) // NOLINT(modernize-loop-convert)
+        {
+            gizmosRenderer.renderDevice->setViewport(viewports[i].position.x, viewports[i].position.y,
+                                                     viewports[i].size.x, viewports[i].size.y);
+            gizmo.draw(gizmosRenderer, mvps[i]);
+        }
+
+        if (gizmo.decreaseLifespan(deltaTime.value))
+        {
+            toRemove.push_back(i);
+        }
+    }
+    while (!toRemove.empty())
+    {
+        std::size_t i = toRemove.back();
+        toRemove.pop_back();
+        gizmos.worldGizmos[i] = gizmos.worldGizmos.back();
+        gizmos.worldGizmos.pop_back();
+    }
+}
+
+static void drawView(Gizmos& gizmos, const ActiveCameras& activeCameras, GizmosRenderer& gizmosRenderer,
+                     const DeltaTime& deltaTime, glm::ivec2 screenSize)
+{
+    if (gizmos.viewGizmos.empty())
+    {
+        return;
+    }
+
+    BaseRenderer::Viewport viewports[4]{};
+
+    int cameraCount = 0;
+
+    for (int i = 0; i < 4; ++i) // NOLINT(modernize-loop-convert)
+    {
+        if (activeCameras.entities[i].isNull())
+        {
+            continue;
+        }
+        cameraCount += 1;
+    }
+
+    splitViewport({0, 0}, screenSize, cameraCount, viewports);
+
+    auto v = glm::translate(glm::mat4(1.0F), glm::vec3(-1.0F, -1.0F, 0.0F));
+    auto p = glm::ortho(-0.5F, 0.5F, -0.5F, 0.5F);
+    auto vp = v * p;
+
+    std::vector<std::size_t> toRemove;
+    for (std::size_t i = 0; i < gizmos.viewGizmos.size(); i++)
+    {
+        auto& gizmo = *(gizmos.viewGizmos[i]);
 
-static void drawGizmosSystem(Write<Gizmos> gizmos, Write<GizmosRenderer> gizmosRenderer, Read<DeltaTime> deltaTime)
+        for (int i = 0; i < cameraCount; ++i) // NOLINT(modernize-loop-convert)
+        {
+            gizmosRenderer.renderDevice->setViewport(viewports[i].position.x, viewports[i].position.y,
+                                                     viewports[i].size.x, viewports[i].size.y);
+            gizmo.draw(gizmosRenderer, vp);
+        }
+
+        if (gizmo.decreaseLifespan(deltaTime.value))
+        {
+            toRemove.push_back(i);
+        }
+    }
+    while (!toRemove.empty())
+    {
+        std::size_t i = toRemove.back();
+        toRemove.pop_back();
+        gizmos.viewGizmos[i] = gizmos.viewGizmos.back();
+        gizmos.viewGizmos.pop_back();
+    }
```

Comment:
Can't you reduce this code duplication with another function? It could be easier if instead of having three vectors you had a single one, and stored the space in the gizmo itself. Then you would just filter them out in each system.

Reply:
I'll be honest, I prefer it this way. The iteration code is repeated, but it's one less thing to be saved on the gizmo itself, it avoids having to do a if else with a whole lot of code within and it let's me just do the mvp math once per space

Reply:
Also, this way it orders them in the screen. viewport is drawn over world, and screen over viewport, which made sense to me